brayns/engine/colormethods/PrimitiveColorMethod.cpp

```diff
@@ -0,0 +1,113 @@
+/* Copyright (c) 2015-2022, EPFL/Blue Brain Project
+ * All rights reserved. Do not distribute without permission.
+ * Responsible Author: Nadir Roman Guerrero <nadir.romanguerrero@epfl.ch>
+ *
+ * This file is part of Brayns <https://github.com/BlueBrain/Brayns>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "PrimitiveColorMethod.h"
+
+#include <brayns/engine/components/ColorList.h>
+#include <brayns/engine/components/ColorSolid.h>
+#include <brayns/engine/components/Geometries.h>
+#include <brayns/engine/components/GeometryViews.h>
+
+#include <brayns/utils/parsing/Parser.h>
+
+namespace
+{
+class ColorListExtractor
+{
+public:
+    static std::vector<brayns::Vector4f> &fromComponents(brayns::Components &components, size_t length)
+    {
+        auto &colorList = components.getOrAdd<brayns::ColorList>();
+        auto &colors = colorList.elements;
+
+        auto solidColor = components.find<brayns::ColorSolid>();
+        if (solidColor)
+        {
+            colors.resize(length, solidColor->color);
+            return colors;
+        }
+
+        colors.resize(length, brayns::Vector4f(1.f));
+        return colors;
+    }
+};
+
+class IndexListExtractor
+{
+public:
+    static std::vector<size_t> fromRange(const std::string &range, size_t maxIndex)
+    {
+        auto indexList = brayns::Parser::extractRanges<size_t>(range);
+        if (indexList.back() >= maxIndex)
+        {
+            throw std::range_error("Range out of bounds: " + range);
+        }
+        return indexList;
+    }
+};
+
+class Painter
+{
+public:
+    static void apply(const std::vector<brayns::Vector4f> &colors, brayns::Components &components)
+    {
+        auto &views = components.get<brayns::GeometryViews>();
+        auto &view = views.elements.front();
+        view.setColorPerPrimitive(ospray::cpp::SharedData(colors));
+        views.modified = true;
+    }
+};
+}
+
+namespace brayns
+{
+PrimitiveColorMethod::PrimitiveColorMethod(std::string name, size_t primitiveCount)
+    : _name(name)
+    , _primitiveCount(primitiveCount)
+{
+}
+
+std::string PrimitiveColorMethod::getName() const
```

Comment:
string_view or is it just called once ?

Reply:
Its called once, but even if it were called more, I need it as a string, so it doesnt makes sense to return a string_view to convert it to string